esphome/components/mcp2515/mcp2515.cpp

```diff
@@ -0,0 +1,641 @@
+    return false;
+  this->set_bitrate_(this->bit_rate_, this->mcp_clock_);
+  this->set_mode_(this->mcp_mode_);
+  ESP_LOGD(TAG, "setup done send test message");
+  return true;
+}
+
+canbus::Error MCP2515::reset_(void) {
+  this->enable();
+  this->transfer_byte(INSTRUCTION_RESET);
+  this->disable();
+  ESP_LOGD(TAG, "reset_()");
+  delay(10);
+
+  ESP_LOGD(TAG, "reset() CLEAR ALL TXB registers");
+
+  uint8_t zeros[14];
+  memset(zeros, 0, sizeof(zeros));
+  set_registers_(MCP_TXB0CTRL, zeros, 14);
+  set_registers_(MCP_TXB1CTRL, zeros, 14);
+  set_registers_(MCP_TXB2CTRL, zeros, 14);
+  ESP_LOGD(TAG, "reset() CLEARED TXB registers");
+
+  set_register_(MCP_RXB0CTRL, 0);
+  set_register_(MCP_RXB1CTRL, 0);
+
+  set_register_(MCP_CANINTE,
+                CANINTF_RX0IF | CANINTF_RX1IF | CANINTF_ERRIF | CANINTF_MERRF);
+
+  modify_register_(MCP_RXB0CTRL, RXBnCTRL_RXM_MASK | RXB0CTRL_BUKT,
+                   RXBnCTRL_RXM_STDEXT | RXB0CTRL_BUKT);
+  modify_register_(MCP_RXB1CTRL, RXBnCTRL_RXM_MASK, RXBnCTRL_RXM_STDEXT);
+
+  return canbus::ERROR_OK;
+}
+
+uint8_t MCP2515::read_register_(const REGISTER reg) {
+  this->enable();
+  this->transfer_byte(INSTRUCTION_READ);
+  this->transfer_byte(reg);
+  uint8_t ret = this->transfer_byte(0x00);
+  this->disable();
+
+  return ret;
+}
+
+void MCP2515::read_registers_(const REGISTER reg, uint8_t values[],
+                              const uint8_t n) {
+  this->enable();
+  this->transfer_byte(INSTRUCTION_READ);
+  this->transfer_byte(reg);
+  // this->transfer_array(values, n);
+  // mcp2515 has auto - increment of address - pointer
+  for (uint8_t i = 0; i < n; i++) {
+    values[i] = this->transfer_byte(0x00);
+  }
+  this->disable();
+}
+
+void MCP2515::set_register_(const REGISTER reg, const uint8_t value) {
+  this->enable();
+  this->transfer_byte(INSTRUCTION_WRITE);
+  this->transfer_byte(reg);
+  this->transfer_byte(value);
+  this->disable();
+}
+
+void MCP2515::set_registers_(const REGISTER reg, uint8_t values[],
+                             const uint8_t n) {
+  this->enable();
+  this->transfer_byte(INSTRUCTION_WRITE);
+  this->transfer_byte(reg);
+  // this->transfer_array(values, n);
+  for (uint8_t i = 0; i < n; i++) {
+    this->transfer_byte(values[i]);
+  }
+  this->disable();
+}
+
+void MCP2515::modify_register_(const REGISTER reg, const uint8_t mask,
+                               const uint8_t data) {
+  this->enable();
+  this->transfer_byte(INSTRUCTION_BITMOD);
+  this->transfer_byte(reg);
+  this->transfer_byte(mask);
+  this->transfer_byte(data);
+  this->disable();
+}
+
+uint8_t MCP2515::get_status_(void) {
+  this->enable();
+  this->transfer_byte(INSTRUCTION_READ_STATUS);
+  uint8_t i = this->transfer_byte(0x00);
+  this->disable();
+
+  return i;
+}
+
+canbus::Error MCP2515::set_mode_(const CANCTRL_REQOP_MODE mode) {
+  modify_register_(MCP_CANCTRL, CANCTRL_REQOP, mode);
+
+  unsigned long endTime = millis() + 10;
+  bool modeMatch = false;
```

Comment:
```suggestion
  bool mode_match = false;
```

See contributing guide and travis checks.